### PR TITLE
Migrate flutter_manifest to null safety

### PR DESCRIPTION
**Pull request by jmagman**
Part of #71511

### CODE REVIEW

**Comment by jmagman** [comment]
This library has a lot of casts, so a close review would be appreciated.

File: packages/flutter_tools/lib/src/flutter_manifest.dart [hunk 1]
```diff
@@ -181,29 +172,29 @@ class FlutterManifest {
...
     }
     return null;
   }
 
   /// Returns the deferred components configuration if declared. Returns
   /// null if no deferred components are declared.
-  List<DeferredComponent> get deferredComponents => _deferredComponents ??= computeDeferredComponents();
-  List<DeferredComponent> _deferredComponents;
-  List<DeferredComponent> computeDeferredComponents() {
+  List<DeferredComponent>? get deferredComponents => _deferredComponents ??= computeDeferredComponents();
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/flutter_manifest.dart:195]: This is probably fine, but its a little weird to do the `_deferredComponents ??= ` trick on a field that could be nullable. Instead, you could make this `late final` (which lacks many of the problems of regular late:

```
late final deferredComponents  = computeDeferredComponents();
```

Then `computeDeferredComponents` will only ever run once, even if it is null

**Review by jonahwilliams — COMMENTED** [review]
You might be able to simplify some of the casting or improve the safety by changing all usage of `dynamic` to `Object?`.

**Review by jonahwilliams — COMMENTED** [review]
👻

**Comment by jmagman** [comment]
> You might be able to simplify some of the casting or improve the safety by changing all usage of `dynamic` to `Object?`.

We're at the mercy of `yaml` package signatures for this file, we'll get cast errors like https://github.com/flutter/flutter/pull/80011 on the collections.  Let me take another look and see if there's anything I can spot that looks safe.

**Comment by jonahwilliams** [comment]
That is if you use `Object`, but `Object?` should be fine.

File: packages/flutter_tools/lib/src/flutter_manifest.dart (hunk 1 above)
Comment (jmagman) [reply packages/flutter_tools/lib/src/flutter_manifest.dart:195]: Oh right, `late final`.  Good call.

**Comment by jmagman** [comment]
> Please get at least one approved review 

Sorry, I forgot this wasn't approved yet.  Thanks bot bro!

**Review by jonahwilliams — APPROVED** [review]
LGTM!

**Pull request merged**